src/main/java/com/google/devtools/build/lib/skyframe/ProcessPackageDirectory.java

```diff
@@ -187,6 +178,35 @@ public ProcessPackageDirectoryResult getPackageExistenceAndSubdirDeps(
         /*additionalValuesToAggregate=*/ ImmutableMap.of());
   }
 
+  // Note that it's not enough to just check for the convenience symlinks themselves,
+  // because if the value of --symlink_prefix changes, the old symlinks are left in place. It
+  // is also not sufficient to check whether the symlink points to a directory in the current
+  // exec root, since this can change between bazel invocations. Therefore we check if the
+  // suffix of the symlink source suggests it is a convenience symlink, then see if the symlink
+  // target is in a directory that looks like an execroot. This algorithm also covers more
+  // creative use cases where people create convenience symlinks somewhere in the directory
+  // tree manually.
+  private boolean isConvenienceSymlink(FileValue fileValue, PathFragment rootRelativePath) {
+    if (!fileValue.isSymlink()) {
+      return false;
+    }
+
+    PathFragment linkTarget = fileValue.getUnresolvedLinkTarget();
+
+    return linkTarget.startsWith(directories.getExecRootBase().asFragment())
+      || (rootRelativePath.getBaseName().endsWith("-bin") && isInExecRoot(linkTarget, 4))
+      || (rootRelativePath.getBaseName().endsWith("-out") && isInExecRoot(linkTarget, 2))
+      || (rootRelativePath.getBaseName().endsWith("-testlogs") && isInExecRoot(linkTarget, 4))
+      || (rootRelativePath.getBaseName().endsWith("-" + directories.getWorkingDirectory().getBaseName()) && isInExecRoot(linkTarget, 1));
+  }
+
+  // TODO: Look at DO_NOT_BUILD_HERE file.
+  private boolean isInExecRoot(PathFragment directory, int depth) {
```

Comment:
What I'd do is the following: pass a `SkyFunction.Environment` here, create a `PathFragment` for the `DO_NOT_BUILD_HERE` file, then do

```
Root absoluteRoot = Root.absoluteRoot(rootedPath.getRoot().geFileSystem());
RootedPath doNotBuildPath = RootedPath.toRootedPath(absoluteRoot, <do not build path>);
FileValue doNotBuildValue = env.getValue(FileValue.key(doNotBuildPath));
if (env.valuesMissing()) { 
  // Make getPackageExistenceAndSubdirDeps() return null early
}

return doNoBuildValue.exists();
```

It's a bit of overkill, but why not do it properly? The ugly hack with the roots doesn't really matter because this code path will be hit very, very infrequently.

Reply:
Okay done (except not being able to use `BlazeWorkspace.DO_NOT_BUILD_FILE_NAME` - see comment below).

I originally tried having these return boxed booleans and propagating nulls, but opted for returning false in the case where dependencies are missing and having one `env.valuesMissing()` after calling them.